File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
import sys
	
input = sys.stdin.readline
print = sys.stdout.write

def isSingleColor(v: list(list())):
    l = len(v)
    
    tmp = 0
    for row in v:
        tmp += sum(row)

    if tmp == 0 or tmp == l*l:
        return True
    else:
        return False

def divideVideo(v: list(list())):
    l = len(v)
    n = l // 2

    tmp = [[[1 for _ in range(n)] for _ in range(n)] for _ in range(4)]
    
    for y in range(l):
        for x in range(l):
            color = v[y][x]
            if y < n:
                # 제 2사분면
                if x < n:
                    tmp[0][y][x] = color
                # 제 1사분면
                else:
                    tmp[1][y][x-n] = color
            else:
                # 제 3사분면
                if x < n:
                    tmp[2][y-n][x] = color
                # 제 4사분면
                else:
                    tmp[3][y-n][x-n] = color

    return tmp

def recursion(v: list(list()), n: int):
    if (n == 1 or isSingleColor(v) == True):
        return str(v[0][0])

    subSquare = divideVideo(v)
    
    return f"({recursion(subSquare[0], n//2)}{recursion(subSquare[1], n//2)}{recursion(subSquare[2], n//2)}{recursion(subSquare[3], n//2)})"

N = int(input())

video = [list(map(int, input().rstrip())) for _ in range(N)]

print(f"{recursion(video, N)}\n")

File: 백준/Silver/1992. 쿼드트리/쿼드트리.swift
// isSingleColor: func([[Int]]) -> Bool
// description: A function that returns Boolean based on wheter the colors 
// in a given 2-Dimensional array are all the same color
func isSingleColor(_ v: [[Int]]) -> Bool {
    // firstColor: First color of  "v"
    let firstColor: Int = v[0][0]

    // Traveling the  v, return false if val is different color from the "firstColor"
    for row in v {
        for val in row {
            if val != firstColor {
                return false
            }
        }
    }

    // Return true when iteration completes search to the end.
    return true
}

// recursion: func([[Int]]) -> String
// description: A recursive function that converts a given 2-Dimensional 
// array into a quad tree and returns it as a string
func recursion(_ v: [[Int]]) -> String {
    // n: Number of rows in  v
    let n: Int = v.count

    // Return first color when no longer divisible or is a single color
    if (n == 1 || isSingleColor(v) == true) {
        return String(v[0][0])
    }

    // half: An integer representing index of half in  v
    let half: Int = n/2

    // top, bot: 2-Dimensional array of integers that stores array v divided
    // between top and bottom based on the "half" index
    let top: [[Int]] = Array(v[..<half])
    let bot: [[Int]] = Array(v[half...])

    // topLeft, ..., botRight: Lists that split list v into four based on half
    var topLeft = [[Int]]()
    var topRight = [[Int]]()
    var botLeft = [[Int]]()
    var botRight = [[Int]]()

    // Append to each list
    for i in 0..<half {
        topLeft.append(Array(top[i][0..<half]))
        topRight.append(Array(top[i][half..<n]))
        botLeft.append(Array(bot[i][0..<half]))
        botRight.append(Array(bot[i][half..<n]))
    }
        
    // Recurvise call divided into top-left, top-right, bottom-left, bottom-right zones
    // and add up the results. 
    return "(" +
        recursion(topLeft) +
        recursion(topRight) +
        recursion(botLeft) +
        recursion(botRight) +
    ")"
}

// N: Int (Not used)
// description: size of given video
// range: 1 ≤ N ≤ 64
if let inputOfN: String = readLine(), let _: Int = Int(inputOfN) {
    // video: [[Int]] (mutable)
    // description: Each color of points in given video
    // range: 0, 1
    // length: N*N
    var video = [[Int]]()

    // Append each line to list "video" 
    while let line: String = readLine() {
        video.append( line.compactMap { Int(String($0)) } )
    }
    
    // Print answer
    print(recursion(video))
}